examples/jsm/webxr/XREstimatedLight.js

```diff
@@ -0,0 +1,225 @@
+import {
+	DirectionalLight,
+	Group,
+	LightProbe,
+	WebGLCubeRenderTarget
+} from "../../../build/three.module.js";
+
+class SessionLightProbe {
+
+	constructor( xrLight, renderer, lightProbe, estimationStartCallback ) {
+
+		this.xrLight = xrLight;
+		this.renderer = renderer;
+		this.lightProbe = lightProbe;
+		this.xrWebGLBinding = null;
+		this.estimationStartCallback = estimationStartCallback;
+		this.frameCallback = this.onXRFrame.bind( this );
+
+		const session = renderer.xr.getSession();
+
+		// If the XRWebGLBinding class is available then we can also query an
+		// estimated reflection cube map.
+		if ( 'XRWebGLBinding' in window ) {
+
+			const gl = renderer.getContext();
+
+			// Ensure that we have any extensions needed to use the preferred cube map format.
+			switch ( session.preferredReflectionFormat ) {
+
+				case 'srgba8':
+					gl.getExtension( 'EXT_sRGB' );
+					break;
+
+				case 'rgba16f':
+					gl.getExtension( 'OES_texture_half_float' );
+					break;
+
+			}
+
+			this.xrWebGLBinding = new XRWebGLBinding( session, gl );
+
+			// TODO: This should actually use the 'reflectionchange' event instead
+			// but there is a known issue in Chrome with that at the moment.
+			/*this.lightProbe.addEventListener('reflectionchange', () => {
+				this.updateReflection();
+			});*/
+
+			// Until the above is fixed, update the reflection once a second. (Which is
+			// approximately how frequently the event would fire anyway).
+			setInterval( () => {
+
+				this.updateReflection();
+
+			}, 1000 );
+
+		}
+
+		// Start monitoring the XR animation frame loop to look for lighting
+		// estimation changes.
+		session.requestAnimationFrame( this.frameCallback );
+
+	}
+
+	updateReflection() {
+
+		const textureProperties = this.renderer.properties.get( this.xrLight.environment );
+
+		if ( textureProperties ) {
+
+			const cubeMap = this.xrWebGLBinding.getReflectionCubeMap( this.lightProbe );
+
+			if ( cubeMap ) {
+
+				textureProperties.__webglTexture = cubeMap;
+
+			}
+
+		}
+
+	}
+
+	onXRFrame( time, xrFrame ) {
```

Comment:
Does this mean the ARCore updates the directional light and SH on every frame but only updates the environment roughly every second? That seems odd; I would have expected all of this to be in the slow update function.

Reply:
ARCore probably only updates those values once in a while, but updating these values is pretty low-overhead (it's just changing which values will be set for a couple of uniforms, which happens per-frame anyway in most cases.) As a result on the API side we didn't consider it a burden to just treat the values as if they're new each frame. We could add an event to indicate concretely when changes happen if needed, but I'd like to see some real-world uses before determining if that's useful or not.

The environment maps are a different story, since not only do they involve creating a texture but also may also involve a behind-the-scenes format conversion. As such they're much heavier weight, relatively, to produce and we structured the API to encourage a sparser query.